File: Week_1/김영범/폰켓몬.js
/*
문제 주소: https://school.programmers.co.kr/learn/courses/30/lessons/1845

풀이 접근법
- nums 로 부터 만들어 질 수 있는 조합의 수를 구하는 것이 아님!
- 전체 몇 종류가 있는지 확인 한다음, 그거 다 가져갈 수 있냐 없냐만 확인하면 된다.
*/

function solution(nums) {
  let answer = 0;

  // 내장 Set 을 이용해 간단히 중복 제거. 이를 통해 주어진 폰켓몬 종 수를 구함.
  const cntForSpecies = new Set(nums).size;
  // 그런데, 어차피 가져갈 수 있는 수 제한은 최대 nums / 2 만큼의 수임.
  const cntForMaxSpecies = nums.length / 2;

  // 조금 더 간단한 풀이, answer = Math.min(cntForSpecies, cntForMaxSpecies)
  answer = cntForSpecies <= cntForMaxSpecies ? cntForSpecies : cntForMaxSpecies;

  return answer;
}

File: Week_1_해시/김성준/폰켓몬.py
# 1. 문제이해
#   총 N마리 폰켓몬 중에서 N/2마리를 선택해야 함
#   선택한 폰켓몬의 종류가 최대한 다양해야 함
#   return: 선택한 폰켓몬의 최대 종류 수

# 2. 예제이해
#   1). nums = [3,3,3,2,2,4]
#       result = 3
#       고유한 폰켓몬의 종류 수  = 중복을 제외하면 {3, 2, 4} = 3
#       선택 가능한 최대 폰켓몬 수 (n/2) = 6/2 = 3
#   2). nums = [3,3,3,2,2,2]
#        result = 2
#       고유한 폰켓몬의 종류 수  = 중복을 제외하면 {3, 2} = 2
#       선택 가능한 최대 폰켓몬 수 (n/2) = 6/2 = 3

# 3. 아이디어 접목
#   1). 중복을 제거해야함
#   2). (n/2)의 개수를 파악
#   3). 두가지의 경우를 전부 다 고려해야 함
#   4). 두가지 경우의 수 중에 최솟값이 정답이다.

# https://school.programmers.co.kr/learn/courses/30/lessons/1845

def solution(nums):

    poketmons = set(nums) # 중복 제거
    maxPoketmons = len(nums) // 2 # n/2의 개수 파악
    return min(len(poketmons), maxPoketmons) # 두가지 경우의 수 중에 최솟값을 구함



